### PR TITLE
[enhancement](invert index) read columns by index reduce seek time

### PR DESCRIPTION
**Pull request by zzzxl1993**
## Proposed changes

Issue Number: close #xxx

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by zzzxl1993** [comment]
run build all

**Comment by zzzxl1993** [comment]
run buildall

**Comment by zzzxl1993** [comment]
run p0

**Comment by zzzxl1993** [comment]
run p0

**Comment by zzzxl1993** [comment]
run p0

**Comment by zzzxl1993** [comment]
run p0

**Comment by zzzxl1993** [comment]
run compile

**Comment by zzzxl1993** [comment]
run p0

**Comment by zzzxl1993** [comment]
run buildall

**Comment by zzzxl1993** [comment]
run buildall

**Comment by zzzxl1993** [comment]
run buildall

**Comment by zzzxl1993** [comment]
run buildall

**Review by airborne12 — APPROVED** [review]
LGTM

File: be/src/olap/rowset/segment_v2/segment_iterator.cpp [hunk 1]
```diff
@@ -1622,8 +1618,49 @@ Status SegmentIterator::_read_columns_by_index(uint32_t nrows_read_limit, uint32
...
+
+    if (_opts.runtime_state) {
+        int32_t threshold = _opts.runtime_state->query_options().inverted_index_read_mode_threshold;
+        if (threshold == 0 || _read_range_info.ranges.size() <= threshold) {
+            RETURN_IF_ERROR(_read_continuous_columns_data(_read_range_info.ranges));
+        } else {
+            RETURN_IF_ERROR(_read_many_columns_data(_read_range_info.rowids));
+        }
+    } else {
+        RETURN_IF_ERROR(_read_continuous_columns_data(_read_range_info.ranges));
```
Comment (xiaokang) [inline be/src/olap/rowset/segment_v2/segment_iterator.cpp:1637]: init threshold to 0 and reduce this branch

File: be/src/olap/rowset/segment_v2/segment_iterator.cpp [hunk 2]
```diff
@@ -1622,8 +1618,49 @@ Status SegmentIterator::_read_columns_by_index(uint32_t nrows_read_limit, uint32
...
+
+        _read_range_info.add(range_from, range_to);
     } while (nrows_read < nrows_read_limit && !_opts.read_orderby_key_reverse);
+
+    if (_read_range_info.ranges.empty()) {
+        return Status::OK();
+    }
+
+    if (_opts.runtime_state) {
+        int32_t threshold = _opts.runtime_state->query_options().inverted_index_read_mode_threshold;
```
Comment (xiaokang) [inline be/src/olap/rowset/segment_v2/segment_iterator.cpp:1630]: the name inverted_index_read_mode_threshold has inverted index, but the logic is not only for inverted index.

File: fe/fe-core/src/main/java/org/apache/doris/qe/SessionVariable.java [hunk 3]
```diff
@@ -1224,6 +1226,12 @@ public void setMaxJoinNumberOfReorder(int maxJoinNumberOfReorder) {
             flag = VariableMgr.GLOBAL)
     public boolean enableFullAutoAnalyze = true;
 
+    @VariableMgr.VarAttr(name = INVERTED_INDEX_READ_MODE_THRESHOLD,
+            description = {"在通过索引查询的时候，根据查询范围的松散程度使用不同的查询方式。",
+                    "When querying through indexes,"
+                    + " use different query methods according to the looseness of the query scope."})
```
Comment (xiaokang) [inline fe/fe-core/src/main/java/org/apache/doris/qe/SessionVariable.java:1232]: sparse

File: be/src/olap/rowset/segment_v2/binary_dict_page.cpp [hunk 4]
```diff
@@ -283,6 +283,7 @@ Status BinaryDictPageDecoder::next_batch(size_t* n, vectorized::MutableColumnPtr
 Status BinaryDictPageDecoder::read_by_rowids(const rowid_t* rowids, ordinal_t page_first_ordinal,
                                              size_t* n, vectorized::MutableColumnPtr& dst) {
     if (_encoding_type == PLAIN_ENCODING) {
+        dst = dst->convert_to_predicate_column_if_dictionary();
```
Comment (xiaokang) [inline be/src/olap/rowset/segment_v2/binary_dict_page.cpp:286]: why

File: be/src/olap/rowset/segment_v2/segment_iterator.h [hunk 5]
```diff
@@ -99,6 +99,28 @@ struct ColumnPredicateInfo {
     std::string query_op;
 };
 
+struct ReadRangeInfo {
+    std::vector<uint32_t> rowids;
+    std::vector<std::pair<uint32_t, uint32_t>> ranges;
+
+    void init(int32_t capacity) {
```
Comment (xiaokang) [inline be/src/olap/rowset/segment_v2/segment_iterator.h:106]: reserve

File: be/src/olap/rowset/segment_v2/segment_iterator.h [hunk 6]
```diff
@@ -99,6 +99,28 @@ struct ColumnPredicateInfo {
     std::string query_op;
 };
 
+struct ReadRangeInfo {
+    std::vector<uint32_t> rowids;
```
Comment (xiaokang) [inline be/src/olap/rowset/segment_v2/segment_iterator.h:103]: It may be very large if many rows selected.

File: be/src/olap/rowset/segment_v2/segment_iterator.h [hunk 7]
```diff
@@ -212,6 +234,9 @@ class SegmentIterator : public RowwiseIterator {
                                        vectorized::MutableColumns& column_block, size_t nrows);
     [[nodiscard]] Status _read_columns_by_index(uint32_t nrows_read_limit, uint32_t& nrows_read,
                                                 bool set_block_rowid);
+    [[nodiscard]] Status _read_continuous_columns_data(
```
Comment (xiaokang) [inline be/src/olap/rowset/segment_v2/segment_iterator.h:237]: the name is not intuitive

File: be/src/vec/columns/column_fixed_length_object.h [hunk 8]
```diff
@@ -286,6 +286,23 @@ class ColumnFixedLengthObject final : public COWHelper<IColumn, ColumnFixedLengt
         memcpy(_data.data() + old_size, data + begin_offset, total_mem_size);
     }
 
+    void insert_many_binary_data(char* data_array, uint32_t* len_array,
```
Comment (xiaokang) [inline be/src/vec/columns/column_fixed_length_object.h:289]: Is this function related to this pr?

**Comment by zzzxl1993** [comment]
run build all

**Comment by zzzxl1993** [comment]
run buildall

File: be/src/olap/rowset/segment_v2/binary_dict_page.cpp (hunk 4 above)
Comment (xiaokang) [reply be/src/olap/rowset/segment_v2/binary_dict_page.cpp:286]: old bug

**Review by xiaokang — APPROVED** [review]
LGTM

**Review by qidaye — APPROVED** [review]
LGTM
